File: realm-browser/src/main/java/de/jonasrottmann/realmbrowser/models/ModelsDiffUtilsCallback.java
package de.jonasrottmann.realmbrowser.models;

import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.annotation.RestrictTo;
import android.support.v7.util.DiffUtil;

import java.util.List;

@RestrictTo(RestrictTo.Scope.LIBRARY_GROUP)
class ModelsDiffUtilsCallback extends DiffUtil.Callback {

    public static final String KEY_CLASS = "KEY_CLASS";
    public static final String KEY_COUNT = "KEY_COUNT";

    private List<ModelPojo> mOldList;
    private List<ModelPojo> mNewList;

    ModelsDiffUtilsCallback(List<ModelPojo> oldList, List<ModelPojo> newList) {
        this.mOldList = oldList;
        this.mNewList = newList;
    }

    @Override
    public int getOldListSize() {
        return mOldList != null ? mOldList.size() : 0;
    }

    @Override
    public int getNewListSize() {
        return mNewList != null ? mNewList.size() : 0;
    }

    @Override
    public boolean areItemsTheSame(int oldItemPosition, int newItemPosition) {
        return mNewList.get(newItemPosition).getKlass().equals(mOldList.get(oldItemPosition).getKlass());
    }

    @Override
    public boolean areContentsTheSame(int oldItemPosition, int newItemPosition) {
        return mNewList.get(newItemPosition).equals(mOldList.get(oldItemPosition));
    }

    @Nullable
    @Override
    public Object getChangePayload(int oldItemPosition, int newItemPosition) {
        ModelPojo newModel = mNewList.get(newItemPosition);
        ModelPojo oldModel = mOldList.get(oldItemPosition);
        Bundle diffBundle = new Bundle();
        if (!newModel.getKlass().equals(oldModel.getKlass())) {
            diffBundle.putSerializable(KEY_CLASS, newModel.getKlass());
        }
        if (newModel.getCount() != oldModel.getCount()) {
            diffBundle.putLong(KEY_COUNT, newModel.getCount());
        }
        if (diffBundle.size() == 0) {
            return null;
        }
        return diffBundle;
    }
}

File: realm-browser/src/main/java/de/jonasrottmann/realmbrowser/models/ModelsDiffUtilsCallback.kt
package de.jonasrottmann.realmbrowser.models

import android.os.Bundle
import android.support.annotation.RestrictTo
import android.support.v7.util.DiffUtil

@RestrictTo(RestrictTo.Scope.LIBRARY_GROUP)
internal class ModelsDiffUtilsCallback(private val mOldList: List<ModelPojo>, private val mNewList: List<ModelPojo>) : DiffUtil.Callback() {

    override fun getOldListSize(): Int = mOldList.size
    override fun getNewListSize(): Int = mNewList.size

    override fun areItemsTheSame(oldItemPosition: Int, newItemPosition: Int): Boolean = mNewList[newItemPosition].klass == mOldList[oldItemPosition].klass
    override fun areContentsTheSame(oldItemPosition: Int, newItemPosition: Int): Boolean = mNewList[newItemPosition] == mOldList[oldItemPosition]

    override fun getChangePayload(oldItemPosition: Int, newItemPosition: Int): Any? {
        val (klassNew, countNew) = mNewList[newItemPosition]
        val (klassOld, countOld) = mOldList[oldItemPosition]

        val diffBundle = Bundle()
        if (klassNew != klassOld) diffBundle.putSerializable(KEY_CLASS, klassNew)
        if (countNew != countOld) diffBundle.putLong(KEY_COUNT, countNew)

        return if (diffBundle.size() == 0) null else diffBundle
    }

    companion object {
        val KEY_CLASS = "KEY_CLASS"
        val KEY_COUNT = "KEY_COUNT"
    }
}
